clang/test/Frontend/sycl-aspects-usage.cpp

```diff
@@ -0,0 +1,68 @@
+// Test checks Propagate Aspect Usage pass.
+
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm %s -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS
+
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown %s -debug-info-kind=constructor -dwarf-version=5 -emit-llvm -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS-DBG -DPATH=%s
+
+// CHECK-WARNINGS: warning: function 'checkStructUsesAspect(int)' uses aspect 'fp16' not listed in 'sycl::device_has()'
+// CHECK-WARNINGS-NEXT: note: the actual use is in checkStructUsesAspect(int), compile with '-g' to get source location
+//
+// CHECK-WARNINGS-DBG: warning: function 'checkStructUsesAspect(int)' uses aspect 'fp16' not listed in 'sycl::device_has()'
+// CHECK-WARNINGS-DBG-NEXT: note: the actual use is in checkStructUsesAspect(int) at [[PATH]]:25:5
+
+#include "Inputs/sycl.hpp"
+
+using namespace cl::sycl;
+
+class KernelName;
+
+struct [[__sycl_detail__::__uses_aspects__(aspect::fp16)]] Struct {
+  int a = 0;
+};
+
+[[sycl::device_has(aspect::fp64)]] int checkStructUsesAspect(int) {
+  Struct s;
+  s.a = 1;
+  return s.a;
+}
+
+[[sycl::device_has()]] int checkEmptyDeviceHas() {
+  return 0;
```

Comment:
So if function `F1` with a set of declared aspects `A` calls function `F2` with a set of actually used aspects `B`. And `B` might not be inside of `A`. Let's `C` call a set `B-A`. Currently, only one arbitrary element from `C` would be printed. This is not user-friendly since a user would need to run compilation several times to figure out the whole set `C`.
We could print one call chain for every missed aspect from `C`. On the other hand, the size of output might blow up and the output will contain lots of repetitions and it might be difficult to extract additional information with comparison to the current way when we print a call chain for only one member of set `C`. Although I am little bit skeptical about printing everything, I am open to objections.

Reply:
I apologize. Checked one more time and currently for every missed aspect from `C` would be printed corresponding call chain.

Reply:
Sorry for not being clear. I was trying to say we need to test this case - 

```
[[sycl::device_has()]] int checkEmptyDeviceHas() {
   funcWithStruct();
}
```

Diagnostic should be generated since we have an empty aspect list.

We also need to test this case - 

```
[[sycl::device_has(aspect::fp16)]] int checkStructUsesAspect2( ) {
  funcWithStruct();
}

[[sycl::device_has()]] int checkEmptyDeviceHas() {
   checkStructUsesAspect2();
}
```

It is not clear to me what the required behavior is here. `checkStructUsesAspect2 `has attribute with `aspect::16`, but it's parent caller has empty aspect list. What shoudl requried behavior be here @gmlueck ?

Reply:
> ```
> [[sycl::device_has()]] int checkEmptyDeviceHas() {
>    funcWithStruct();
> }
> ```
> 
> Diagnostic should be generated since we have an empty aspect list.

Just to make sure we are on the same page, it is perfectly OK to have an empty aspect list in `[[sycl::device_has()]]`.  This means "my function doesn't require a device that has any specific aspects".  If such a function *does* use a feature that requires an aspect, then the compiler  should raise a diagnostic.  In your example above, does `funcWithStruct()` use a feature that requires an aspect?  Is this why you say that a diagnostic should be generated?


> We also need to test this case -
> 
> ```
> [[sycl::device_has(aspect::fp16)]] int checkStructUsesAspect2( ) {
>   funcWithStruct();
> }
> 
> [[sycl::device_has()]] int checkEmptyDeviceHas() {
>    checkStructUsesAspect2();
> }
> ```
> 
> It is not clear to me what the required behavior is here. `checkStructUsesAspect2 `has attribute with `aspect::16`, but it's parent caller has empty aspect list. What shoudl requried behavior be here @gmlueck ?

This should cause a diagnostic.  The `[[sycl::device_has(aspect::fp16)]]` on `checkStructUsesAspect2` is an assertion that this function requires that aspect.  See this sentence in SYCL 2020:

>When this attribute appears in a translation unit that calls the decorated device function, it is an assertion that the device function uses optional features that correspond to the aspects listed in the attribute.

The `[[sycl::device_has()]]` on `checkEmptyDeviceHas` asks the compiler to generate a diagnostic if that function uses any feature that requires any aspect (because the aspect list is empty).  Since `checkStructUsesAspect2` requires `fp16`, a diagnostic should be generated.

Reply:
>does funcWithStruct() use a feature that requires an aspect? Is this why you say that a diagnostic should be generated?

Yes `funcWithStruct()` uses a type marked with `aspect::fp16`

Thanks for the explanation. That makes sense. @maksimsab could you please add these tests to ensure behavior is as expected?

Reply:
Fixed `checkEmptyDeviceHas()` and added `checkEmptyDeviceHas2()` with your new example.